File: products/llm_analytics/frontend/text-view/formatters/messageFormatter.ts
/**
 * Format input and output messages for text view
 */
import { isObject } from 'lib/utils'

interface ToolCall {
    function?: {
        name: string
        arguments: string | Record<string, any>
    }
    name?: string
    args?: string | Record<string, any>
}

interface Message {
    role?: string
    type?: string
    content?: string | any[]
    tool_calls?: ToolCall[]
}

/**
 * Truncate content with middle ellipsis for long text
 * Embeds full content in a special marker format for click-to-expand functionality
 */
function truncateContent(content: string, maxLength = 1000): { lines: string[]; truncated: boolean } {
    if (content.length <= maxLength) {
        return { lines: [content], truncated: false }
    }

    const half = Math.floor(maxLength / 2)
    const firstPart = content.slice(0, half)
    const lastPart = content.slice(-half)
    const truncatedChars = content.length - maxLength
    const middlePart = content.slice(half, -half)

    // Encode the middle part in a special marker format that we'll parse in the display component
    // Format: <<<TRUNCATED|base64content|charCount>>>
    const encodedMiddle = btoa(encodeURIComponent(middlePart))
    const marker = `<<<TRUNCATED|${encodedMiddle}|${truncatedChars}>>>`

    return {
        lines: [firstPart, '', marker, '', lastPart],
        truncated: true,
    }
}

/**
 * Format tool calls for display
 */
function formatToolCalls(toolCalls: ToolCall[]): string[] {
    const lines: string[] = []
    lines.push(`Tool calls: ${toolCalls.length}`)

    for (const tc of toolCalls) {
        // Handle both OpenAI format (function: {name, arguments})
        // and LangChain format (name, args)
        let name: string
        let args: string | Record<string, any>

        if (tc.function) {
            name = tc.function.name
            args = tc.function.arguments
        } else {
            name = tc.name || 'unknown'
            args = tc.args || ''
        }

        // Parse args into object if needed
        let parsedArgs: Record<string, any> | null = null
        if (typeof args === 'object') {
            parsedArgs = args
        } else if (typeof args === 'string' && args) {
            try {
                parsedArgs = JSON.parse(args)
            } catch {
                // If parsing fails, will show raw string
            }
        }

        // Format as function call
        if (parsedArgs && typeof parsedArgs === 'object') {
            const argEntries = Object.entries(parsedArgs)
            if (argEntries.length > 1) {
                // Multiline for multiple args
                lines.push(`  - ${name}(`)
                for (const [k, v] of argEntries) {
                    lines.push(`      ${k}=${JSON.stringify(v)},`)
                }
                lines.push('    )')
            } else if (argEntries.length === 1) {
                // Single line for one arg
                const [k, v] = argEntries[0]
                lines.push(`  - ${name}(${k}=${JSON.stringify(v)})`)
            } else {
                // No args
                lines.push(`  - ${name}()`)
            }
        } else {
            // Fallback for unparseable args
            lines.push(`  - ${name}(${args || ''})`)
        }
    }

    return lines
}

/**
 * Extract text content from various message content formats
 */
function extractTextContent(content: any): string {
    if (typeof content === 'string') {
        return content
    }

    if (Array.isArray(content)) {
        const textParts: string[] = []
        for (const block of content) {
            if (isObject(block)) {
                if ('text' in block) {
                    textParts.push(block.text)
                } else if ('type' in block && block.type === 'tool_use') {
                    textParts.push(`[Tool use: ${block.name || 'unknown'}]`)
                }
            } else {
                textParts.push(String(block))
            }
        }
        return textParts.join('\n')
    }

    return String(content)
}

/**
 * Format input messages section
 */
export function formatInputMessages(aiInput: any): string[] {
    const lines: string[] = []

    if (!aiInput || (Array.isArray(aiInput) && aiInput.length === 0)) {
        return lines
    }

    lines.push('')
    lines.push('INPUT:')

    // Handle simple string input
    if (typeof aiInput === 'string') {
        lines.push('')
        lines.push('[User input]')
        const { lines: contentLines } = truncateContent(aiInput)
        lines.push(...contentLines)
        return lines
    }

    // Handle array of message objects
    if (Array.isArray(aiInput)) {
        for (let i = 0; i < aiInput.length; i++) {
            const msg = aiInput[i] as Message
            const role = msg.role || msg.type || 'unknown'
            const content = msg.content || ''
            const toolCalls = msg.tool_calls || []

            lines.push('')
            lines.push(`[${i + 1}] ${role.toUpperCase()}`)
            lines.push('')

            if (content) {
                const textContent = extractTextContent(content)
                if (textContent) {
                    const { lines: contentLines } = truncateContent(textContent)
                    lines.push(...contentLines)
                }
            }

            if (toolCalls.length > 0) {
                lines.push('')
                lines.push(...formatToolCalls(toolCalls))
            }

            // Add separator between messages (but not after the last one)
            if (i < aiInput.length - 1) {
                lines.push('')
                lines.push('-'.repeat(80))
            }
        }
        return lines
    }

    // Unknown format - show raw
    lines.push('')
    lines.push(`[Unparsed input format: ${typeof aiInput}]`)
    lines.push(JSON.stringify(aiInput).slice(0, 500))

    return lines
}

/**
 * Format output messages section
 */
export function formatOutputMessages(aiOutput: any, aiOutputChoices: any): string[] {
    const lines: string[] = []

    // Simple string output
    if (aiOutput && typeof aiOutput === 'string') {
        lines.push('')
        lines.push('OUTPUT:')
        lines.push('')
        const { lines: contentLines } = truncateContent(aiOutput)
        lines.push(...contentLines)
        return lines
    }

    // Extract choices array if wrapped in an object (e.g., xai format: {choices: [...]})
    let choices = aiOutputChoices
    if (aiOutputChoices && typeof aiOutputChoices === 'object' && !Array.isArray(aiOutputChoices)) {
        if ('choices' in aiOutputChoices && Array.isArray(aiOutputChoices.choices)) {
            choices = aiOutputChoices.choices
        }
    }

    // Output choices (most common format)
    if (choices && Array.isArray(choices) && choices.length > 0) {
        lines.push('')
        lines.push('OUTPUT:')

        for (let i = 0; i < choices.length; i++) {
            const choice = choices[i]

            // Handle different output choice formats
            let role: string
            let content: any
            let toolCalls: ToolCall[] = []

            if (typeof choice === 'string') {
                role = 'assistant'
                content = choice
            } else if ('message' in choice) {
                // Nested message format
                const msg = choice.message
                role = msg.role || msg.type || 'unknown'
                content = msg.content || ''
                toolCalls = msg.tool_calls || []
            } else {
                role = choice.role || choice.type || 'unknown'
                content = choice.content || ''
                toolCalls = choice.tool_calls || []
            }

            // Label tool messages specially
            const roleLabel = role.toLowerCase() === 'tool' ? 'TOOL RESULT' : role.toUpperCase()

            lines.push('')
            lines.push(`[${i + 1}] ${roleLabel}`)
            lines.push('')

            if (content || content === '') {
                const textContent = extractTextContent(content)
                if (textContent) {
                    const { lines: contentLines } = truncateContent(textContent)
                    lines.push(...contentLines)
                } else if (!toolCalls.length) {
                    lines.push('[empty response]')
                }
            }

            if (toolCalls.length > 0) {
                lines.push('')
                lines.push(...formatToolCalls(toolCalls))
            }

            // Add separator between messages (but not after the last one)
            if (i < choices.length - 1) {
                lines.push('')
                lines.push('-'.repeat(80))
            }
        }
    }

    return lines
}

File: products/llm_analytics/backend/text_repr/formatters/message_formatter.py
"""
Format input and output messages for text view.

Handles formatting of LLM messages including role-based conversations, tool calls,
and content blocks. Supports multiple provider formats (OpenAI, Anthropic, etc.)
with truncation and interactive markers for frontend display.
"""

import json
import base64
from typing import Any, TypedDict
from urllib.parse import quote

from .constants import DEFAULT_TRUNCATE_BUFFER, MAX_UNABLE_TO_PARSE_REPR_LENGTH, MAX_UNPARSED_DISPLAY_LENGTH


class FormatterOptions(TypedDict, total=False):
    """Options for formatting text representations."""

    truncated: bool  # Use truncation for long content (default: True)
    truncate_buffer: int  # Chars to show at start/end (default: 1000)
    include_markers: bool  # Use interactive markers vs plain text (default: True)
    collapsed: bool  # Show full hierarchy vs summary (default: False)


class ToolCall(TypedDict, total=False):
    """Tool call structure supporting multiple formats."""

    function: dict[str, Any]  # OpenAI format: {name, arguments}
    name: str  # LangChain format
    args: Any  # LangChain format


class Message(TypedDict, total=False):
    """Message structure."""

    role: str
    type: str
    content: Any
    tool_calls: list[ToolCall]


def truncate_content(content: str, options: FormatterOptions | None = None) -> tuple[list[str], bool]:
    """
    Truncate content with middle ellipsis for long text.
    Can use interactive markers (for frontend) or plain text indicators (for backend/LLM).

    Returns:
        Tuple of (lines, truncated_flag)
    """
    if options is None:
        options = {}

    should_truncate = options.get("truncated", True)  # Default: True
    max_length = options.get("truncate_buffer", DEFAULT_TRUNCATE_BUFFER)
    use_markers = options.get("include_markers", True)  # Default: True

    if not should_truncate or len(content) <= max_length:
        return ([content], False)

    half = max_length // 2
    first_part = content[:half]
    last_part = content[-half:]
    truncated_chars = len(content) - max_length

    if use_markers:
        # Frontend: encoded marker for expand/collapse UI
        middle_part = content[half:-half]
        # Base64 encode the URL-encoded middle part
        encoded_middle = base64.b64encode(quote(middle_part).encode()).decode()
        marker = f"<<<TRUNCATED|{encoded_middle}|{truncated_chars}>>>"
        return ([first_part, "", marker, "", last_part], True)
    else:
        # Backend: plain text indicator for LLM context
        marker = f"\n\n... ({truncated_chars} chars truncated) ...\n\n"
        return ([first_part + marker + last_part], True)


def format_single_tool_call(name: str, args: Any) -> str:
    """Format a single tool call as a function signature."""
    # Parse args into dict if needed
    parsed_args: dict[str, Any] | None = None

    if isinstance(args, dict):
        parsed_args = args
    elif isinstance(args, str) and args:
        try:
            parsed_args = json.loads(args)
        except json.JSONDecodeError:
            # If parsing fails, will show raw string
            pass

    # Format as function call
    if parsed_args and isinstance(parsed_args, dict):
        arg_entries = list(parsed_args.items())
        if arg_entries:
            arg_strings = [f"{k}={json.dumps(v)}" for k, v in arg_entries]
            return f"{name}({', '.join(arg_strings)})"
        return f"{name}()"
    elif args:
        # Fallback for unparseable args
        return f"{name}({args})"
    return f"{name}()"


def format_tool_calls(tool_calls: list[ToolCall]) -> list[str]:
    """Format tool calls for display."""
    lines: list[str] = []
    lines.append(f"Tool calls: {len(tool_calls)}")

    for tc in tool_calls:
        # Handle both OpenAI format (function: {name, arguments})
        # and LangChain format (name, args)
        if tc.get("function"):
            name = tc["function"].get("name", "unknown")
            args = tc["function"].get("arguments", "")
        else:
            name = tc.get("name", "unknown")
            args = tc.get("args", "")

        lines.append(f"  - {format_single_tool_call(name, args)}")

    return lines


def extract_tool_calls_from_content(content: Any) -> list[ToolCall]:
    """Extract tool calls from content array."""
    if not isinstance(content, list):
        return []

    tool_calls: list[ToolCall] = []
    for block in content:
        if isinstance(block, dict):
            # Handle tool-call format: { type: "tool-call", function: {...} }
            if block.get("type") == "tool-call" and "function" in block:
                if isinstance(block["function"], dict):
                    tool_calls.append({"function": block["function"]})
            # Handle Anthropic function format: { type: "function", function: {...} }
            elif block.get("type") == "function" and "function" in block:
                if isinstance(block["function"], dict):
                    tool_calls.append({"function": block["function"]})

    return tool_calls


def safe_extract_text(content: Any) -> str:
    """
    Safely extract text from various content formats.
    Handles strings, dicts with 'text' key, arrays of content blocks, etc.

    Includes data preservation in error cases to aid debugging and discovery of edge cases.
    """
    try:
        if isinstance(content, str):
            return content

        if isinstance(content, dict):
            # Try common text keys
            if "text" in content:
                return str(content["text"])
            if "content" in content:
                return safe_extract_text(content["content"])
            # Fallback to JSON representation
            return json.dumps(content)

        if isinstance(content, list):
            text_parts = []
            for i, item in enumerate(content):
                if isinstance(item, dict):
                    item_type = item.get("type")
                    # Try both "text" and "content" keys (tool_result uses "content")
                    text_value = item.get("text") or item.get("content")

                    if text_value:
                        # Add spacing between all content blocks in the array
                        if i > 0 and text_parts:
                            text_parts.append("")  # Blank line separator

                        # If we have a type and it's not just "text", label it
                        if item_type and item_type != "text":
                            text_parts.append(f"[{item_type.upper()}]")
                            text_parts.append("")  # Blank line after label

                        text_parts.append(str(text_value))
                elif isinstance(item, str):
                    # Add spacing before string items too
                    if i > 0 and text_parts:
                        text_parts.append("")
                    text_parts.append(item)
            if text_parts:
                return "\n".join(text_parts)

        # Fallback with data preservation for debugging
        data_repr = repr(content)[:MAX_UNABLE_TO_PARSE_REPR_LENGTH]
        if len(repr(content)) > MAX_UNABLE_TO_PARSE_REPR_LENGTH:
            data_repr += "..."
        return f"[UNABLE_TO_PARSE: {type(content).__name__}] {data_repr}"
    except Exception as e:
        # Handle any unexpected errors during extraction
        try:
            data_repr = repr(content)[:MAX_UNABLE_TO_PARSE_REPR_LENGTH]
            return f"[PARSE_ERROR: {type(content).__name__}] {str(e)} | Data: {data_repr}"
        except:
            return f"[PARSE_ERROR: {type(content).__name__}] {str(e)}"


def _is_special_block(block: Any) -> bool:
    """Check if a block needs special handling (tool calls, functions, etc)."""
    if not isinstance(block, dict):
        return False

    block_type = block.get("type")
    if block_type in ("tool-call", "tool_use", "function"):
        return True

    # Check for tool-call content format
    if "content" in block and isinstance(block["content"], dict) and "toolName" in block["content"]:
        return True

    return False


def _format_special_block(block: dict) -> str | None:
    """
    Format special blocks (tool calls, tool_use, etc).

    Returns:
        Formatted string if block was handled, None if block should be skipped.
    """
    block_type = block.get("type")

    # Handle tool-call type directly (format: {type: "tool-call", toolName, input})
    if block_type == "tool-call":
        tool_name = block.get("toolName", "unknown")
        tool_input = block.get("input", {})
        return format_single_tool_call(tool_name, tool_input)

    # Handle tool_use type (Anthropic format)
    if block_type == "tool_use":
        tool_name = block.get("name", "unknown")
        tool_input = block.get("input", {})

        # If input is empty, check for partial_json field
        if not tool_input and "partial_json" in block:
            try:
                tool_input = json.loads(block["partial_json"])
            except (json.JSONDecodeError, ValueError):
                pass

        return format_single_tool_call(tool_name, tool_input)

    # Skip function blocks as they'll be handled separately
    if block_type == "function":
        return None

    # Handle tool-call content for inline display
    if "content" in block:
        block_content = block["content"]
        if isinstance(block_content, dict) and "toolName" in block_content:
            tool_name = block_content.get("toolName", "unknown")
            args = block_content.get("args", "")
            return format_single_tool_call(tool_name, args)
        # Handle tool-result content
        elif isinstance(block_content, dict) and "result" in block_content:
            tool_name = block_content.get("toolName", "unknown")
            return f"[Tool result: {tool_name}]"

    # Block not handled by this function
    return safe_extract_text(block)


def extract_text_content(content: Any) -> str:
    """
    Extract text content from various message content formats.
    Uses safe extraction with fallback for unparseable content.

    Handles special blocks like tool calls inline for better readability.
    """
    # Handle special cases that need inline formatting (tool calls, etc)
    if isinstance(content, list):
        # Check if any blocks need special handling
        if any(_is_special_block(block) for block in content):
            text_parts: list[str] = []
            for block in content:
                if isinstance(block, dict):
                    formatted = _format_special_block(block)
                    if formatted is not None:
                        # Skip empty strings and UNABLE_TO_PARSE markers from non-text blocks
                        if formatted and not formatted.startswith("[UNABLE_TO_PARSE"):
                            text_parts.append(formatted)
                # Handle non-dict items in list
                elif isinstance(block, str):
                    text_parts.append(block)

            if text_parts:
                return "\n\n".join(text_parts)

    # Use safe extraction for non-special content (handles type labels for text/reasoning/etc)
    return safe_extract_text(content)


def format_input_messages(ai_input: Any, options: FormatterOptions | None = None) -> list[str]:
    """Format input messages section."""
    lines: list[str] = []

    if not ai_input or (isinstance(ai_input, list) and len(ai_input) == 0):
        return lines

    lines.append("")
    lines.append("INPUT:")

    # Handle simple string input
    if isinstance(ai_input, str):
        lines.append("")
        content_lines, _ = truncate_content(ai_input, options)
        lines.extend(content_lines)
        return lines

    # Handle array of message objects
    if isinstance(ai_input, list):
        for i, msg in enumerate(ai_input):
            if not isinstance(msg, dict):
                continue

            role = msg.get("role") or msg.get("type") or "unknown"
            content = msg.get("content", "")
            tool_calls = msg.get("tool_calls", [])

            lines.append("")
            lines.append(f"[{i + 1}] {role.upper()}")
            lines.append("")

            if content:
                text_content = extract_text_content(content)
                if text_content:
                    content_lines, _ = truncate_content(text_content, options)
                    lines.extend(content_lines)

            if tool_calls:
                lines.append("")
                lines.extend(format_tool_calls(tool_calls))

            # Add separator between messages (but not after the last one)
            if i < len(ai_input) - 1:
                lines.append("")
                lines.append("-" * 80)

        return lines

    # Unknown format - show raw with data preservation
    lines.append("")
    lines.append(f"[Unparsed input format: {type(ai_input).__name__}]")
    json_repr = json.dumps(ai_input)
    if len(json_repr) > MAX_UNPARSED_DISPLAY_LENGTH:
        json_repr = json_repr[:MAX_UNPARSED_DISPLAY_LENGTH] + "..."
    lines.append(json_repr)

    return lines


def format_output_messages(
    ai_output: Any, ai_output_choices: Any, options: FormatterOptions | None = None
) -> list[str]:
    """Format output messages section."""
    lines: list[str] = []

    # Simple string output
    if ai_output and isinstance(ai_output, str):
        lines.append("")
        lines.append("OUTPUT:")
        lines.append("")
        content_lines, _ = truncate_content(ai_output, options)
        lines.extend(content_lines)
        return lines

    # Extract choices array if wrapped in an object (e.g., xai format: {choices: [...]})
    choices = ai_output_choices
    if ai_output_choices and isinstance(ai_output_choices, dict) and "choices" in ai_output_choices:
        if isinstance(ai_output_choices["choices"], list):
            choices = ai_output_choices["choices"]

    # Output choices (most common format)
    if choices and isinstance(choices, list) and len(choices) > 0:
        lines.append("")
        lines.append("OUTPUT:")

        for i, choice in enumerate(choices):
            if not isinstance(choice, dict):
                continue

            # Extract message from choice
            # Handle both OpenAI format (choice.message) and Anthropic format (choice is the message)
            message = choice.get("message")
            if not message or not isinstance(message, dict):
                # Anthropic/direct format - choice IS the message
                if "role" in choice or "content" in choice:
                    message = choice
                else:
                    continue

            role = message.get("role", "assistant")
            content = message.get("content", "")
            tool_calls = message.get("tool_calls", [])

            # Extract tool calls from content if present
            content_tool_calls = extract_tool_calls_from_content(content)
            if content_tool_calls:
                tool_calls = content_tool_calls

            lines.append("")
            lines.append(f"[{i + 1}] {role.upper()}")
            lines.append("")

            if content:
                text_content = extract_text_content(content)
                if text_content:
                    content_lines, _ = truncate_content(text_content, options)
                    lines.extend(content_lines)

            if tool_calls:
                lines.append("")
                lines.extend(format_tool_calls(tool_calls))

        return lines

    # Fallback - no recognizable output format
    return lines
